### PR TITLE
Propose seccomp/apparmor protobuf type definitions for CRI graduation

### PR DESCRIPTION
**Pull request by saschagrunert**
**What type of PR is this?**

/kind design

**What this PR does / why we need it**:
With the graduation of the Container Runtime Interface (CRI) we will have a unique chance to change the protobuf protocol. We now use native types for seccomp and AppArmor to make the API more clear.

**Which issue(s) this PR fixes**:
None

**Special notes for your reviewer**:
cc @mrunalp @mikebrow @pjbgf 

**Does this PR introduce a user-facing change?**:

```release-note
None
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs
None
```

### CODE REVIEW

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto [hunk 1]
```diff
@@ -279,13 +279,29 @@ message LinuxSandboxSecurityContext {
     // This allows a sandbox to take additional security precautions if no
     // privileged containers are expected to be run.
     bool privileged = 6;
-    // Seccomp profile for the sandbox, candidate values are:
-    // * runtime/default: the default profile for the container runtime
-    // * unconfined: unconfined profile, ie, no seccomp sandboxing
-    // * localhost/<full-path-to-profile>: the profile installed on the node.
-    //   <full-path-to-profile> is the full path of the profile.
-    // Default: "", which is identical with unconfined.
-    string seccomp_profile_path = 7;
```
Comment (mrunalp) [inline staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto:288]: We need to do this in 2 steps so container runtimes could be updated to use the new field.

Comment (saschagrunert) [reply staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto:288]: Yes that's why the field is now deprecated. I think this just implies a `Deprecated: Do not use` comment in the actual go code but then linters can hint to it.

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto [hunk 2]
```diff
@@ -285,7 +285,30 @@ message LinuxSandboxSecurityContext {
...
+        // A profile defined in a file on the node should be used.
+        Localhost = 2;
+    }
+    // Indicator which `ProfileType` should be applied.
+    ProfileType profile_type = 1;
+    // Indicates a profile defined in a file on the node should be used. The
+    // profile must be preconfigured on the node to work. Must be a descending
+    // path, relative to the kubelet's configured seccomp profile location.
+    // Must only be set if `ProfileType` is `Localhost`.
+    string localhost_profile = 2;
```
Comment (mrunalp) [inline staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto:311]: Should we call this `localhost_profile_path`?

**Comment by mrunalp** [comment]
@saschagrunert Static check is failing on deprecation:
```
Errors from staticcheck:
pkg/kubelet/dockershim/docker_container.go:183:42: config.GetLinux().GetSecurityContext().GetSeccompProfilePath is deprecated: Do not use.  (SA1019)
```

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto (hunk 2 above)
Comment (saschagrunert) [reply staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto]: Fixed as suggested.

**Comment by saschagrunert** [comment]
> @saschagrunert Static check is failing on deprecation:
> 
> ```
> Errors from staticcheck:
> pkg/kubelet/dockershim/docker_container.go:183:42: config.GetLinux().GetSecurityContext().GetSeccompProfilePath is deprecated: Do not use.  (SA1019)
> ```

Fixed by excluding from the staticcheck.

File: hack/.staticcheck_failures [hunk 3]
```diff
@@ -1,6 +1,7 @@
 cluster/images/etcd/migrate
 pkg/controller/podautoscaler
 pkg/controller/replicaset
+pkg/kubelet/dockershim
```
Comment (dims) [inline hack/.staticcheck_failures:4]: unrelated change?

Comment (saschagrunert) [reply hack/.staticcheck_failures:4]: No the `deprecated` field will cause staticcheck to fail in that file. That's why I excluded it here.

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto [hunk 4]
```diff
@@ -644,7 +667,10 @@ message LinuxContainerSecurityContext {
     // * localhost/<full-path-to-profile>: the profile installed on the node.
     //   <full-path-to-profile> is the full path of the profile.
     // Default: "", which is identical with unconfined.
-    string seccomp_profile_path = 10;
+    string seccomp_profile_path = 10 [deprecated=true];
+    // Seccomp profile for the contaner, defaults to the `RuntimeDefault`
```
Comment (mrunalp) [inline staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto:671]: typo: contaner -> container

Comment (mrunalp) [inline staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto:671]: Can we defer the default to `RuntimeDefault` after we merge in the CRI changes?

Comment (saschagrunert) [reply staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto:671]: Fixed :+1:

Comment (saschagrunert) [reply staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto:671]: We can, I would prefer doing it now. We can work around in the runtimes and act accordingly to break no tests. :)

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto [hunk 5]
```diff
@@ -285,7 +285,30 @@ message LinuxSandboxSecurityContext {
     // * localhost/<full-path-to-profile>: the profile installed on the node.
     //   <full-path-to-profile> is the full path of the profile.
     // Default: "", which is identical with unconfined.
-    string seccomp_profile_path = 7;
+    string seccomp_profile_path = 7 [deprecated=true];
+    // Seccomp profile for the sandbox, defaults to the `RuntimeDefault`
```
Comment (mrunalp) [inline staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto:289]: This comment needs to be updated as well. Thanks!

Comment (saschagrunert) [reply staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto]: Done :+1:

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto [hunk 6]
```diff
@@ -285,7 +285,29 @@ message LinuxSandboxSecurityContext {
...
+        // A profile defined in a file on the node should be used.
+        Localhost = 2;
+    }
+    // Indicator which `ProfileType` should be applied.
+    ProfileType profile_type = 1;
+    // Indicates a profile defined in a file on the node should be used. The
+    // profile must be preconfigured on the node to work. Must be a descending
+    // path, relative to the kubelet's configured seccomp profile location.
+    // Must only be set if `ProfileType` is `Localhost`.
+    string localhost_profile_path = 2;
```
Comment (mikebrow) [inline staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto:310]: per discussion let's go with full path

Comment (saschagrunert) [reply staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto:310]: Fixed :+1:

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto [hunk 7]
```diff
@@ -279,13 +279,37 @@ message LinuxSandboxSecurityContext {
     // This allows a sandbox to take additional security precautions if no
     // privileged containers are expected to be run.
     bool privileged = 6;
+    // Seccomp profile for the sandbox.
+    SecurityProfile seccomp_profile = 9;
+    // AppArmor profile for the sandbox.
+    SecurityProfile apparmor_profile = 10;
```
Comment (saschagrunert) [inline staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto:285]: @mikebrow @mrunalp PTAL, this is a slightly different Approach where we can re-use the same type `SecurityProfile` for seccomp and AppArmor.

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto [hunk 8]
```diff
@@ -631,29 +655,33 @@ message LinuxContainerSecurityContext {
...
+    // Seccomp profile for the container.
+    SecurityProfile seccomp_profile = 15;
+    // AppArmor profile for the container.
+    SecurityProfile appamor_profile = 16;
     // AppArmor profile for the container, candidate values are:
     // * runtime/default: equivalent to not specifying a profile.
     // * unconfined: no profiles are loaded
     // * localhost/<profile_name>: profile loaded on the node
     //    (localhost) by name. The possible profile names are detailed at
     //    http://wiki.apparmor.net/index.php/AppArmor_Core_Policy_Reference
```
Comment (mikebrow) [inline staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto:676]: link is broken..

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto [hunk 9]
```diff
@@ -631,29 +655,33 @@ message LinuxContainerSecurityContext {
...
+    // masked_paths is a slice of paths that should be masked by the container
+    // runtime, this can be passed directly to the OCI spec.
+    repeated string masked_paths = 13;
+    // readonly_paths is a slice of paths that should be set as readonly by the
+    // container runtime, this can be passed directly to the OCI spec.
+    repeated string readonly_paths = 14;
+    // Seccomp profile for the container.
+    SecurityProfile seccomp_profile = 15;
+    // AppArmor profile for the container.
+    SecurityProfile appamor_profile = 16;
```
Comment (mikebrow) [inline staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto:670]: consider just "apparmor"

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto [hunk 10]
```diff
@@ -631,29 +655,33 @@ message LinuxContainerSecurityContext {
...
+    // container.
+    bool no_new_privs = 11;
+    // masked_paths is a slice of paths that should be masked by the container
+    // runtime, this can be passed directly to the OCI spec.
+    repeated string masked_paths = 13;
+    // readonly_paths is a slice of paths that should be set as readonly by the
+    // container runtime, this can be passed directly to the OCI spec.
+    repeated string readonly_paths = 14;
+    // Seccomp profile for the container.
+    SecurityProfile seccomp_profile = 15;
```
Comment (mikebrow) [inline staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto:668]: consider just seccomp

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto [hunk 11]
```diff
@@ -279,13 +279,37 @@ message LinuxSandboxSecurityContext {
...
+        // A pre-defined profile on the node should be used.
+        Localhost = 2;
+    }
+    // Indicator which `ProfileType` should be applied.
+    ProfileType profile_type = 1;
+    // Indicates that a pre-defined profile on the node should be used. The
+    // profile must be preconfigured on the node to work. Must be an absolute
+    // path if the profile references a file. Must only be set if `ProfileType`
+    // is `Localhost`.
+    string localhost_profile_path = 2;
```
Comment (mikebrow) [inline staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto:312]: suggest localhost_ref..
and mention that for apparmor...
and for seccomp...

**Review by mikebrow — COMMENTED** [review]
see naming comments

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto (hunk 9 above)
Comment (saschagrunert) [reply staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto]: Fixed :+1:

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto (hunk 10 above)
Comment (saschagrunert) [reply staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto]: Fixed :+1:

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto (hunk 11 above)
Comment (saschagrunert) [reply staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto]: Fixed :+1:

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto (hunk 8 above)
Comment (saschagrunert) [reply staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto]: Fixed :+1:

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto [hunk 12]
```diff
@@ -279,13 +279,37 @@ message LinuxSandboxSecurityContext {
...
+        Unconfined = 1;
+        // A pre-defined profile on the node should be used.
+        Localhost = 2;
+    }
+    // Indicator which `ProfileType` should be applied.
+    ProfileType profile_type = 1;
+    // Indicates that a pre-defined profile on the node should be used.
+    // Must only be set if `ProfileType` is `Localhost`.
+    // For seccomp, it must be an absolute path to the profile.
+    // For AppArmor, it must be equal to the loaded profile name.
```
Comment (mikebrow) [inline staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto:311]: s/it must be equal to the loaded profile name/this field is the AppArmor `<profile name>`/

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto [hunk 13]
```diff
@@ -279,13 +279,37 @@ message LinuxSandboxSecurityContext {
...
+        // Disable the feature for the sandbox or the container.
+        Unconfined = 1;
+        // A pre-defined profile on the node should be used.
+        Localhost = 2;
+    }
+    // Indicator which `ProfileType` should be applied.
+    ProfileType profile_type = 1;
+    // Indicates that a pre-defined profile on the node should be used.
+    // Must only be set if `ProfileType` is `Localhost`.
+    // For seccomp, it must be an absolute path to the profile.
```
Comment (mikebrow) [inline staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto:310]: s/to the/to the seccomp/

**Review by mikebrow — COMMENTED** [review]
couple comments on the comments :-)

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto (hunk 12 above)
Comment (saschagrunert) [reply staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto]: Fixed as suggested :+1:

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto (hunk 13 above)
Comment (saschagrunert) [reply staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto]: Fixed as suggested :+1:

**Comment by saschagrunert** [comment]
> couple comments on the comments :-)

Thanks! Addressed them too :+1:

**Review by mikebrow — APPROVED** [review]
LGTM

**Comment by mrunalp** [comment]
@dchen1107 ptal

**Review by derekwaynecarr — APPROVED** [review]
changes look good.

/lgtm
/approve
